Common++/header/LRUList.h

```diff
@@ -72,11 +76,52 @@ namespace pcpp
 			return NULL;
 		}
 
+		/**
+		 * Puts an element in the list. This element will be inserted (or advanced if it already exists) to the head of the
+		 * list as the most recently used element. If the list already reached its max size and the element is new this method
+		 * will remove the least recently used element and return a value in deletedValue. Method complexity is O(log(getSize())).
+		 * This is a optimized version of the method T* put(const T&).
+		 * @param[in] element The element to insert or to advance to the head of the list (if already exists)
+		 * @param[out] deletedValue The value of deleted element (if a pointer is not NULL)
+		 * @return 0 if the list didn't reach its max size, 1 otherwise. In case the list already reached its max size
+		 * and deletedValue is not NULL the value of deleted element is copied into the place the deletedValue points to.
+		 */
+		int put(const T& element, T* deletedValue)
+		{
+			m_CacheItemsList.push_front(element);
+
+			// Inserting a new element. If an element with an equivalent key already exists the method returns an iterator to the element that prevented the insertion
+			std::pair<MapIterator, bool> pair = m_CacheItemsMap.insert(std::make_pair(element, m_CacheItemsList.begin()));
+			if (pair.second == false) // already exists
+			{
+				m_CacheItemsList.erase(pair.first->second);
+				pair.first->second = m_CacheItemsList.begin();
+			}
+
+			if (m_CacheItemsMap.size() > m_MaxSize)
+			{
+				ListIterator lruIter = m_CacheItemsList.end();
+				lruIter--;
+
+				if (deletedValue != NULL)
+#if __cplusplus > 199711L || _MSC_VER >= 1800
+					*deletedValue = std::move(*lruIter);
+#else
+					*deletedValue = *lruIter;
+#endif
```

Comment:
This is a nice optimization! However I don't think we need 2 versions of `put()`, we should choose between the two. This version is more optimized but the API is a bit less readable IMO. What do you think?

Reply:
I suggest to remove the old method and to change the new one to following:

`int put(const T& element, T* deletedValue = NULL)`

The user should only make changes to process the new type of return value (`deletedValue` has the default value).

Reply:
Yes I agree. Please notice `LRUList` is also used in 2 examples: `PcapSplitter` and `TcpReassembly`. Please make sure that you make the changes there too. In addition, I'd really appreciate if you can run some tests on these examples after the change to make sure they behave correctly

Reply:
The changes have been made. I have run the tests with following parameters and had the same results for different versions of certain application:

- `TcpReassembly -r three_http_streams.pcap -f 2`

- `PcapSplitter -f sll.pcap -o old -m client-ip -p 5`

In this case I have changed the `MAX_NUMBER_OF_CONCURRENT_OPEN_FILES` constant to 5 before compilation.

In addition, `valgrind` has found the memory leakages in examples.